issue_9202 / minimum function ALC

baiti opened the pull request:
Putting it into a drafts PR such that others can review it and let me know then whether I am moving into the right direction with it.

Renders dots on the map. In details view some more data is shown e.g. owner and when published. Labs app can be launched via Open in Browser. No further details. No description of the adventure, no waypoints at all.

Tomorrow I will add a launch capability in details similar to  WIGs and Chirps. Further, I will make it PMO ;(

bekuno commented on main/src/cgeo/geocaching/enumerations/CacheType.java:
```diff
@@ -36,6 +36,7 @@
     GPS_EXHIBIT("gps", "GPS Adventures Exhibit", "72e69af2-7986-4990-afd9-bc16cbbb4ce3", R.string.gps, R.drawable.type_event, "1304", R.drawable.dot_event), // icon missing
     BLOCK_PARTY("block", "Geocaching HQ Block Party", "bc2f3df2-1aab-4601-b2ff-b5091f6c02e3", R.string.block, R.drawable.type_event, "4738", R.drawable.dot_event), // icon missing
     LOCATIONLESS("locationless", "Locationless (Reverse) Cache", "8f6dd7bc-ff39-4997-bd2e-225a0d2adf9d", R.string.locationless, R.drawable.type_virtual, "12", R.drawable.dot_virtual), // icon missing
+    ADVLAB("advlab", "Adventure Lab Cache", "", R.string.advlab, R.drawable.type_advlab, "7777", R.drawable.dot_advlab),
```
Official GC Cache types are defined in https://www.geocaching.com/about/cache_types.aspx
Therefore I would stay with `Lab Cache`. 
The `pattern` and `wptTypeId` are used/defined in the cache website. But we have no cache site for lab caches. Is the `wptTypeId` defined/used in the lab cache data? 

This additional row created the CI error 
`cgeo.geocaching.apps.AbstractLocusAppTest > testToLocusTypeCount FAILED`
Please increase the count number in the test.
The Locus app does (currently) not support lab caches, see 
https://github.com/asamm/locus-api/blob/82d0b51472dc71420df0fea23280cb34dc97ee8c/locus-api-core/src/main/java/locus/api/objects/geocaching/GeocachingData.kt#L623

baiti replied:
Adventure Lab Cache -> Lab Cache: fixed

cgeo.geocaching.apps.AbstractLocusAppTest > testToLocusTypeCount FAILED .... increased from 10 to 11, not knowing the logic behind all this at this time but will follow up later.

With respect to `pattern` and `wptTypeId`: No, none of those come down via the API method. Now: what does that mean?

bekuno replied:
> cgeo.geocaching.apps.AbstractLocusAppTest > testToLocusTypeCount FAILED .... increased from 10 to 11, not knowing the logic behind all this at this time but will follow up later.

Fine.
The logic here is to detect that we changed something at the caches types, so it is necessary to check if we need to configure the locus connector. This is ok for now.

> With respect to `pattern` and `wptTypeId`: No, none of those come down via the API method. Now: what does that mean?

Let them blank.

baiti commented:
Codacy complains about an area in the code I haven't  touched:

```
main/src/cgeo/geocaching/CacheDetailActivity.java#L1586
Avoid unused local variables such as 'intent'.
```

Should I go ahead and fix this?

bekuno commented on main/src/cgeo/geocaching/connector/lc/LCConnector.java:
```diff
@@ -89,7 +89,7 @@ public String getHost() {
     public SearchResult searchByGeocode(@Nullable final String geocode, @Nullable final String guid, final DisposableHandler handler) {
-        //Log.e("searchByGeocode: geocode = " + geocode);
+        Log.d("searchByGeocode: geocode = " + geocode);
         if (geocode == null) {
```
`geocode` can be null.

baiti replied:
fixed locally. Will go into next commit

bekuno commented on main/src/cgeo/geocaching/enumerations/CacheType.java:
```diff
@@ -36,7 +36,7 @@
     BLOCK_PARTY("block", "Geocaching HQ Block Party", "bc2f3df2-1aab-4601-b2ff-b5091f6c02e3", R.string.block, R.drawable.type_event, "4738", R.drawable.dot_event), // icon missing
     LOCATIONLESS("locationless", "Locationless (Reverse) Cache", "8f6dd7bc-ff39-4997-bd2e-225a0d2adf9d", R.string.locationless, R.drawable.type_virtual, "12", R.drawable.dot_virtual), // icon missing
-    ADVLAB("advlab", "Adventure Lab Cache", "", R.string.advlab, R.drawable.type_advlab, "7777", R.drawable.dot_advlab),
+    ADVLAB("advlab", "Lab Cache", "", R.string.advlab, R.drawable.type_advlab, "7777", R.drawable.dot_advlab),
```
Do not use a dummy `wptTypeId` (7777).
See e.g. USER_DEFINED cache type.
Maybe it is better to rename all without "Adventure"?

baiti replied:
> Do not use a dummy `wptTypeId` (7777).

> See e.g. USER_DEFINED cache type.
> Maybe it is better to rename all without "Adventure"?

Of course, no problem, just let me know your preference, however,  a global change won't do as GC HQ isn't sure how to call this thing either. The application ist called "Adventure Lab".

Can we come to some recommendations as a team as to exactlty how to name this things in all the various places in the code? I am open to the team's decision.

baiti replied:
Once comment wrt Lab / Adventure Lab: I think GC HQ has rebranded that game. Years ago those were called "Lab Cache" and they were played at Mega / Giga Events. Since about 2 years ago these things are called Adventure Lab Caches. I'd assume they just haven't renamed it system wide (yet)

baiti commented:
> Codacy complains about an area in the code I haven't touched:
> 
> ```
> main/src/cgeo/geocaching/CacheDetailActivity.java#L1586
> Avoid unused local variables such as 'intent'.
> ```
> 
> Should I go ahead and fix this?

Closed (actually I introduced this bug inadvertedly)

Lineflyer commented on main/res/values/strings_not_translatable.xml:
```diff
@@ -188,6 +188,7 @@
     <string translatable="false" name="caches_map_locus">Locus</string>
     <string translatable="false" name="settings_title_gc">Geocaching.com</string>
     <string translatable="false" name="settings_title_ec">Extremcaching.com</string>
+    <string translatable="false" name="settings_title_lc">Labs.geocaching.com</string>
```
Shouldn't we rather call it "Geocaching Adventure Lab" or similar?

moving-bits commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -1591,6 +1592,28 @@ private void updateChirpWolfBox() {
+                final ImageButton buttonALC = view.findViewById(R.id.send_to_alc);
+                buttonALC.setOnClickListener(v -> {
+                    // re-check installation state, might have changed since creating the view
+                    if (null != alc) {
```
Strictly speaking you should do an actual reevaluation whether ALC package is still present (so do a new call to `ProcessUtils.getLaunchIntent`), as the user might have uninstalled it meanwhile. A null check for `alc` won't work here.
OTOH it's not very likely, that a user opens a cache detail page, uninstalls ALC, and the taps the button, so we can probably remove the `if (null != alc)` check completely.

moving-bits replied:
Ignore my previous comment :-(
Have just realized that you also offer a link to the app store. For this you still need the null check.

baiti commented:
From my perspective it is ready for review and eventuall a merge then.

The implementation is now as discussed in the meeting. With respect to "Basic" it is compliant with GC HQ's request. Of course there are multiple implementation options with respect to how to accomplish this compliance.

I left the preferences dialog in place and also the fake login (you can put it any arbitrary username / password there). At the point where we decide we will never consider playing in "active mode" this can be eliminated entirely, the only setting we need is an "Activate" checkbox ... which we may then decide to bind to PMO.

Before we refactor all this I'd see we need to come to terms with respect to "partner API" usage as that could change the concept even more.

Lineflyer commented:
> From my perspective it is ready for review and eventuall a merge then.

Sounds good.
Please take a look to make the checks green. AFAICS its just a minor issue (wrong import order) left over to be fixed.

Lineflyer commented:
Addition: 
Also two locus related tests failing.

baiti commented:
> > From my perspective it is ready for review and eventuall a merge then.
> 
> Sounds good.
> Please take a look to make the checks green. AFAICS its just a minor issue (wrong import order) left over to be fixed.

I don't see Codacy complaining.

moving-bits commented:
> I don't see Codacy complaining.

It's not Codacy this time, but the build process (which also triggers the tests) - have a look at the other entry in "checks", marked with the red cross in front of it.

moving-bits commented:
FYI: There are a couple of potential error sources reported in the build section:

![image](https://user-images.githubusercontent.com/3754370/114444504-4601f100-9bcf-11eb-95b6-93e731c422d4.png)

In this case we have:
- Lint marks 86 founds - but not all of them are new (or even none). We shall not introduce new lint errors.
- CheckStyle marks 1 error - not actively introduced by you in this case, but CheckStyle also triggers existing errors if you touch code in the same method.
- There is one failing test,
- and in this case this is also the reason why in the last section the build aborts.
 
So in this case, fixing the CheckStyle section and the one failing test should do the trick. (Haven't checked whether new lint messages got introduced.)

baiti commented:
It is still complaining. It is the line:

assertEquals(12, CacheSize.values().length);

It was 11 before, I increased it by 1. However, isn't this just trial and error?  it is kind of surprising to me that we have to hardcode some number of elements in a structure. Back in the old days of C we had the "sizeof"  compile time operator. Isn't there something like this in Java?

bekuno commented:
You can run in AS the tests locally.

You added a cache type, but changed the cache size test.

Lineflyer commented:
> This isn't related to this PR but if you want I can try to fix it anyway ;)

I guess its related, because you imported the `LCConnector` in line 14 while it belongs somewhere below the remarked wrong import order.

baiti commented:
> > This isn't related to this PR but if you want I can try to fix it anyway ;)
> 
> I guess its related, because you imported the `LCConnector` in line 14 while it belongs somewhere below the remarked wrong import order.

dammit ... ;) I updated AS to 4.3, now checkstyle is no longer working. Will locate the help I got a while ago on how to configure checkstyle again, then I will fix the "wrong order" and then I will try to the the test thing going. Stay tuned.

UniQP commented on main/src/cgeo/geocaching/connector/lc/LCApi.java:
```diff
@@ -0,0 +1,279 @@
+    private static Geocache importCacheFromJSON(final Response response) {
+        try {
+            final JsonNode json = JsonUtils.reader.readTree(Network.getResponseData(response));
+            //Log.e(json.toPrettyString());
```
Please remove debugging code

baiti replied:
Done

UniQP commented on main/src/cgeo/geocaching/connector/lc/LCApi.java:
```diff
@@ -0,0 +1,279 @@
+            }
+            final List<Geocache> caches = new ArrayList<>(items.size());
+            for (final JsonNode node : items) {
+                //Log.e(node.toPrettyString());
```
Please remove debugging code

baiti replied:
Done

UniQP commented on main/src/cgeo/geocaching/connector/lc/LCApi.java:
```diff
@@ -0,0 +1,279 @@
+        final Geopoint pointZero = new Geopoint(0, 0);
+        int stageCounter = 0;
+        for (final JsonNode wptResponse: wptsJson) {
+            //Log.e(wptResponse.toPrettyString());
```
Please remove debugging code

baiti replied:
Done ... for the sole purpose of being in line with your guidelines. In my own projects I leave those in as I frequently need the  in the next debugging cyle ;)

eddiemuc replied:
@baiti you may use `Log.d` or `Log.v` for debug/verbose logs. Then there would not be the need to comment them out and you can turn them on as needed during development or Bugfix.

baiti replied:
Good to know. Next time then. In order to have some filtering, would I also specify a "TAG"? I remember when I did that Kotlin thing, the  function was `Log.x(TAG, msg);` I assume I could use tag for filtering?

eddiemuc replied:
@baiti cgeo has it's own Log class which does the TAG assignment internally (basically the tag is cgeo). In its methods you can't specify a tag yourself. But of course You can use certain keywords in the log message itself to search for afterwards. I often e.g. put the method name on the log message for this purpose

UniQP commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -1591,6 +1592,28 @@ private void updateChirpWolfBox() {
+            final LinearLayout boxCW = view.findViewById(R.id.alc_box);
+            boxCW.setVisibility(isEnabled ? View.VISIBLE : View.GONE);
+            final TextView msgCW = view.findViewById(R.id.alc_text);
+            msgCW.setText(null != alc ? R.string.cache_alc_start : R.string.cache_alc_install);
```
```suggestion
            msgCW.setText(alc != null ? R.string.cache_alc_start : R.string.cache_alc_install);
```
Personally, I like to have `null` on the right-hand side because "value is not null" reads more natural than "null is not value".

baiti replied:
Me too. I have copy / pasted it from Chrip and I did that for consistency. Now if I change it for ALC, do you also want me to change it for Chirp?

moving-bits replied:
I propose to fix that separately, as it's unrelated to this PR, and there are probably many more place where we have "null" at the left hand side. We should tackle them all in one separate PR (except those "introduced" with this PR, those should be fixed here).

baiti replied:
since I had to commit it for the ALC related part, I fixed Chirp too

UniQP commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -1591,6 +1592,28 @@ private void updateChirpWolfBox() {
+                final ImageButton buttonALC = view.findViewById(R.id.send_to_alc);
+                buttonALC.setOnClickListener(v -> {
+                    // re-check installation state, might have changed since creating the view
+                    if (null != alc) {
```
```suggestion
                    if (alc != null) {
```

(See above.)

UniQP commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -1237,9 +1237,10 @@ public NestedScrollView getDispatchedView(final ViewGroup parentView) {
             updateWatchlistBox();
 
-            // WhereYouGo and ChirpWolf
+            // WhereYouGo and ChirpWolf and Adventure Lab
```
```suggestion
            // WhereYouGo, ChirpWolf and Adventure Lab
```

Lineflyer commented:
Hurray...It's green :)

baiti commented:
> Hurray...It's green :)

Yeah ... but with quite some frustration remaining. AS was bugging me since weeks to update to 4.3. I just ignored it but today I yielded to the temptation. Now my checkstyle is gone and I have not been able to get it back working. I have spent so much time now manouvering and navigaring around endlessly in the AS' UI labyrinth. I know there are quite a few on the team who are also commandline guys like me so I know there will be help to get me there.

git stuff works nice on the commandline now. the "./gradle ....." not yet as I never have used gradle before this project.

bekuno commented:
Considering the new GC challange "THE SCIENCE OF DISCOVERY" https://www.geocaching.com/play/leaderboard I propose to integrate this PR in release after releasing the next c:geo version and then (a week after?) make the next release (bug fix + ALC min).

bekuno commented:
> 
> 
> > > This isn't related to this PR but if you want I can try to fix it anyway ;)
> > 
> > 
> > I guess its related, because you imported the `LCConnector` in line 14 while it belongs somewhere below the remarked wrong import order.
> 
> dammit ... ;) I updated AS to 4.3, now checkstyle is no longer working. Will locate the help I got a while ago on how to configure checkstyle again, then I will fix the "wrong order" and then I will try to the the test thing going. Stay tuned.

There is no AS 4.3, do you mean 4.1.3?
For checkstyle integration in AS you need the checkstyle plugin. Is this installed and updated?
And is your AS checkstyle configuration configured/updated - see https://github.com/cgeo/cgeo/blob/master/.idea/checkstyle-idea.xml ?

bekuno commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -1237,9 +1237,10 @@ public NestedScrollView getDispatchedView(final ViewGroup parentView) {
             updateWatchlistBox();
 
-            // WhereYouGo and ChirpWolf
+            // WhereYouGo,ChirpWolf and Adventure Lab
```
There is a space missing:
`// WhereYouGo, ChirpWolf and Adventure Lab`

baiti replied:
You guys are taking the review really serious. I am exactly the same type of guy ;) Thanks for pointing it out. Will fix with next push.

baiti replied:
commit done

baiti commented:
> There is no AS 4.3, do you mean 4.1.3?

Yes, 4.1.3

> For checkstyle integration in AS you need the checkstyle plugin. Is this installed and updated?

![Screenshot from 2021-04-13 11-31-58](https://user-images.githubusercontent.com/5296333/114531144-eac21400-9c4b-11eb-8868-b4d13c506c83.png)


> And is your AS checkstyle configuration configured/updated - see https://github.com/cgeo/cgeo/blob/master/.idea/checkstyle-idea.xml ? 

Doesn't this come down automatically via `git fetch master`? ... I manually checked and this file is there in my local repo.

I assume it is just that I cannot find the right point to click on in that overwhelming UI of AS.

baiti commented:
> Considering the new GC challange "THE SCIENCE OF DISCOVERY" https://www.geocaching.com/play/leaderboard I propose to integrate this PR in release after releasing the next c:geo version and then (a week after?) make the next release (bug fix + ALC min).

In general I certainly agree, however, the PR needs a bit more work in order to take out the Login / Authorization stuff from the UI.

Actually I have done this already in my local repo and I am now searching for the point where the click in SettingsActivity's "Activate" checkbox actually activates the container. If any one of you could jump in here and give me a hand it would speed up things and the PR would be ready quickly. I am sure it is "just" making one call at the right place in the code.

eddiemuc commented:
> > For checkstyle integration in AS you need the checkstyle plugin. Is this installed and updated?
> 
> ![Screenshot from 2021-04-13 11-31-58](https://user-images.githubusercontent.com/5296333/114531144-eac21400-9c4b-11eb-8868-b4d13c506c83.png)
> 
> > And is your AS checkstyle configuration configured/updated - see https://github.com/cgeo/cgeo/blob/master/.idea/checkstyle-idea.xml ?
> 

@baiti I see from your screenshot that you use version 5.51.0 of checkstyle plugin. I upgraded to this version recently and found that checkstyle plugin no longer works on my AS (problem seems to be that AS runs under Java 8 while plugin requires Java 11). I downgraded again to plugin version 5.48.0, now it works again. Maybe on your installation you have a similar problem?

bekuno commented:
> I assume it is just that I cannot find the right point to click on in that overwhelming UI of AS.

You can run `gradlew check` or `gradlew checkstyle`.

For sorting the imports you can use the AS menu entry "Code -> Optimize Imports", or the key-code CTRL+ALT+O.
In the commit dialog you can activate the checkbutton "Optimize imports".

baiti commented:
> @baiti I see from your screenshot that you use version 5.51.0 of checkstyle plugin. I upgraded to this version recently and found that checkstyle plugin no longer works on my AS (problem seems to be that AS runs under Java 8 while plugin requires Java 11). I downgraded again to plugin version 5.48.0, now it works again. Maybe on your installation you have a similar problem?

@eddiemuc  ...That was **exactly** the problem.

baiti commented:
> > I assume it is just that I cannot find the right point to click on in that overwhelming UI of AS.
> 
> You can run `gradlew check` or `gradlew checkstyle`.
>

I had tried those commands but they didn't work. Now I know why. It was because the checkstyle was also updated to 5.51.0 in the process of updating AS.

> For sorting the imports you can use the AS menu entry "Code -> Optimize Imports", or the key-code CTRL+ALT+O.
> In the commit dialog you can activate the checkbutton "Optimize imports".

Good to know! Thanks a lot.

baiti commented:
I have two topics I am struggling with. Once these are done we could actually merge.

1. Activate / deactivate a connector
2. Show / Hide a PreferenceScreen based on member status

Focusing on (1): I have removed the OAuthPreference from the preference_screen_lc as I do not need any authorization nor a login. I just want the checkbox called "Activate". From a UI perspective this is working and if I toggle the box to unchecked, the connector goes down and is no longer called. Exactly what I wanted.

Now if I go into Settings again and activate that connector it is working at the UI level but the connector stays off more precisely: it is not called for getting caches. If I completely terminate the application, and restart it, it is back on and working. So something must be happening in fresh initalization that I also have to do at the point after the checkbox is checked into "Active" but I have not found out what makes it really active. The "isLoggedIn" and "isActive" methods return true in all cases so it must be something else.

Does anybody have an idea where to look and for what to look?

baiti commented:
While debugging the code in this PR I have noticed strange behaviour, which I would normally write up as an issue, however, I cannot be sure that it is also in the latest official release, which I don't have installed anywhere.

Here is the scenario:
- Log into GC with a Premium account
- Logout of Premium Account
- Log into GC with a Basic account

Whils still in the "Update or remove authorization" screen, the avatar from the previously logged in ID is still shown, further,
the `Options for Basic Members` ist still greyed out.

If I now backout up to the Main screen, then go back into Settings -> Services, the erronous avatar is gone, however, the `Options for Basic Members`  is stil greyed out.

If I now completely kill the application and restart it, `Options for Basic Members` is clickable. But goes on: If I now logout from the Basic acount, then log into my Premium account I will not get the avatar back and th `Options for Basic Members` remains cklickable. If I go one screen back and then back in, the avatar shows, but `Options for Basic Members` remains enabled.

I am sure that I am experiencing the same underlying failures in my own attempts to disable ALC for basic members.

bekuno commented:
This scenario is a known problem.
You can solve it by prior to entering the new users login use the long press on "Authorize c:geo" to remove the existing login data.
See eg. the discussion in #7581.

Lineflyer commented:
I have just checked out this PR and installed it.
Somehow I am not able to get it working:
I do see the connector in Settings - Services and activated it. Still the live map is not showing adventures.
Any work in progress here or should it work?

BTW:
- As suggested as comment to the code I would rather call it "Geocaching Adventure Lab" instead of "labs.geocaching.com"
- I would place the connector right below geocaching.com in the connector list in Settings - Services

baiti commented:
> I have just checked out this PR and installed it.
> Somehow I am not able to get it working:
> I do see the connector in Settings - Services and activated it. Still the live map is not showing adventures.
> Any work in progress here or should it work?

Unfortunately after activating the connectore, you need to terminate the application and start cgeo fresh to make it work.

> 
> BTW:
> 
> * As suggested as comment to the code I would rather call it "Geocaching Adventure Lab" instead of "labs.geocaching.com"

Ok, will do

Could you pls review all text strings related to ALC and make a proposal on the wording in each and every case? That would help me a lot.


> * I would place the connector right below geocaching.com in the connector list in Settings - Services

Ok, will do

fm-sys commented:
> * As suggested as comment to the code I would rather call it "Geocaching Adventure Lab" instead of "labs.geocaching.com"
> 
> * I would place the connector right below geocaching.com in the connector list in Settings - Services

What about placing it as subsetting of Geocaching.com instead?

baiti commented:
> > * As suggested as comment to the code I would rather call it "Geocaching Adventure Lab" instead of "labs.geocaching.com"
> > * I would place the connector right below geocaching.com in the connector list in Settings - Services
> 
> What about placing it as subsetting of Geocaching.com instead?

Sure ... it could be a simple checkbox "Activate Adventure Lab Caches" in the GC preferences Screen, it should become clickable if the logged in user is Premium

Lineflyer commented:
> What about placing it as subsetting of Geocaching.com instead?

I am undecided about this one.
From logical perspective its perfect, but for visibility and separation to our (non API) geocaching.com access and potential future (maybe) evolvement it might just be listed as a dedicated other connector in the list. Not sure...

Lineflyer commented:
Another small remark:
Could we try to beautify the string which is used as geocode?
It runs over two complete lines on my low res devices and its probably anyway nothing a user would note down, forward or use in any means...isn't it?

I remember there was a reference code in the beginning, which the user could even define hisself.
E.g. for my ALC I used `VelAdventure01` and the link to https://labs.geocaching.com/goto/VelAdventure01 is still working. 
However if I look into the ALC details its now refered to as https://adventurelab.page.link/KSQT

Maybe (if this information is available) we could use this as geocode?

BTW: What URL do you use for "Open in browser"?

Lineflyer commented:
Sorry for spamming the PR (but I thought its more useful than opening an issue for something which is not yet merged):

The description page is (logically for now) empty.
I was wondering whether it might be explanatory to fill it with a short explanation text (similar as we do for user defined caches) with something like:
```
At the moment the adventure lab description is available only by using the Adventure Lab app. Go to the details page and install/open the adventure lab app to get more details.
```

baiti commented:
> Another small remark:
> Could we try to beautify the string which is used as geocode?

I have spent **days** to create the illusion of make it look like a GC-code. I never got it working. The whole system works with geocode and I have to put the UUID into that property otherwise I cannot launch into `CacheDetailActivity`. If one of the experts can show me how I could accomplish this change, I would be motivated and willing to do it.

> It runs over two complete lines on my low res devices and its probably anyway nothing a user would note down, forward or use in any means...isn't it?

I am all in. But: I need help for these things. If you just leave it with me, it won't happen. I have tried very hard.
 
> 
> I remember there was a reference code in the beginning, which the user could even define hisself.
> E.g. for my ALC I used `VelAdventure01` and the link to https://labs.geocaching.com/goto/VelAdventure01 is still working.
> However if I look into the ALC details its now refered to as https://adventurelab.page.link/KSQT
> 
> Maybe (if this information is available) we could use this as geocode?
> 
> BTW: What URL do you use for "Open in browser"?

`CACHE_URL = "https://adventurelab.page.link/`

baiti commented:
> Sorry for spamming the PR (but I thought its more useful than opening an issue for something which is not yet merged):
> 
> The description page is (logically for now) empty.
> I was wondering whether it might be helpful to fill it with a short explanation text (similar as we do for user defined caches) with something like:
> 
> "At the moment the adventure lab description is available only by using the Adventure Lab app. Go to the details page and install/open the adventure lab app to get more details."

Piece of cake. Just give me the exact wording you want.

baiti commented:
> > What about placing it as subsetting of Geocaching.com instead?
> 
> I am undecided about this one.
> From logical perspective its perfect, but for visibility and separation to our (non API) geocaching.com access and potential future (maybe) evolvement it might just be listed as a dedicated other connector in the list. Not sure...

Actually a quick question here: Normally discussion boards such as for instance `phpbb` and also many of todays messenger apps provide the capability to setup a **poll**... which we could use to vote for a specific proposal. Do we have such a capability somewhere. You have noticed I am relatively quick with changing things, however, it is not so much fun to change it back and forth and back again ;)

Should I perhaps slow down a bit an wait untill you guys had some time to digest things? I know, you all have day jobs and this is just for fun here. I am probably the only one who is in the luxury position to manage my own time at leisure

Lineflyer commented:
> Should I perhaps slow down a bit and wait untill you guys had some time to digest things?

Maybe the best method would be to first of all finish this PR functionswise and based on code review. Then we merge and track the remaining bugs or improvement proposals in some dedicated issues.  Thats AFAIK the implicit method we used successfully in the past.
So maybe my remarks and proposals have been a bit too early.

baiti commented:
> > Should I perhaps slow down a bit and wait untill you guys had some time to digest things?
> 
> Maybe the best method would be to first of all finish this PR functionswise and based on code review. Then we merge and track the remaining bugs or improvement proposals in some dedicated issues. Thats AFAIK the implicit method we used successfully in the past.

Ok, that sounds like a good strategy. So there is **one** big item that needs to be addressed. It happens to be the one you stumbled over. The one I have reported here earlier today. It is about getting the connector really up and working after the change at in the SettingsActivity has been made. Currently the application must be terminated, not just closed and reopened.

Unfortunately this is a type of a problem I need help. I have spent a full day in the code and have not found out how to address this. Also pls note, other parts in the system have similar problems so it is not just limited to the ALC function.

The fastest, and most efficient way to get this done is in some sort of a "hackathon". It won't help to kick mails back and forth at a rate of 1 or 2 per day. We need to set aside some quality time, let's say two hours in jitsi and then get it done together. We look at the code together. Make the changes together. Two hours with a guy who knows his way around in Android and all its APIs

Once we have fixed this, I think we can carefully review and inspect the code a s a preparation for the merge.


> So maybe my remarks and proposals have been a bit too early.

naaaah ... all good ;)

baiti commented:
> > Another small remark:
> > Could we try to beautify the string which is used as geocode?

I have given this some thoughts: A geocode, for the regular game, is something a geocacher can remember and it frequently comes up that companion geoacher asks hey, can you give me this geocode quickly, or that I punch it manually into some search field in whatever other application. A geocode for Adventure Lab Caches is completely irrelevant for the geocacher to ever see or deal with. This is true for the UUID as well as the shorter patten that is used when launching the app.

So here is one possibility: We could go into the code where that code is rendered and completely eliminate this information from the views if that code starts with "LC". Thoughts?

baiti commented:
How about this:

![Screenshot from 2021-04-15 10-15-50](https://user-images.githubusercontent.com/5296333/114837215-c2652180-9dd3-11eb-8414-e9490bd4bb84.png)

fm-sys commented:
> It is about getting the connector really up and working after the change at in the SettingsActivity has been made. Currently the application must be terminated, not just closed and reopened.

I think there's something like `SettingsActivity.setResult(RESTART_NEEDED)` (don't remember it exactly). You can look it up at the AppLanguageSelector in the settings. AFAIK I had used it there...

baiti commented:
> > It is about getting the connector really up and working after the change at in the SettingsActivity has been made. Currently the application must be terminated, not just closed and reopened.
> 
> I think there's something like `SettingsActivity.setResult(RESTART_NEEDED)` (don't remember it exactly). You can look it up at the AppLanguageSelector in the settings. AFAIK I had used it there...

Yes, and changing the one and only occurance of NO_RESTART_NEEDED to RESTART_NEEDED, makes it actually work. I will go and try to find the place where the .setResult(RESTART_NEEDED) must be added....

baiti commented:
> > > It is about getting the connector really up and working after the change at in the SettingsActivity has been made. Currently the application must be terminated, not just closed and reopened.
> > 
> > 
> > I think there's something like `SettingsActivity.setResult(RESTART_NEEDED)` (don't remember it exactly). You can look it up at the AppLanguageSelector in the settings. AFAIK I had used it there...
> 
> Yes, and changing the one and only occurance of NO_RESTART_NEEDED to RESTART_NEEDED, makes it actually work. I will go and try to find the place where the .setResult(RESTART_NEEDED) must be added....

I have it working! .... The screen flip looks a bit ugly but it is indeed working. Would be interesting to see what exactly it is in the RESTART path that makes ist working and then take perhaps that single statement and put it where I now have the .setResult()

baiti commented:
I have added a single line of code into `SettingsActivity` and now Jenkins compalins all over the place :-(

Other than that. With this restart thing now addressed (thanks to @fm-sys ), I'd say strictly from a functional perspective we could review and merge now.

What remains to be done, and what can be done later as optimizations:

- Come up with agreed upon terminology for everything related to ALC
- Shorten the geocode in ListViews and on the Popup that shows up when a dot on the map is clicked. So far I have been unable to figure out where exactly the data binding is done otherwise I would have fixed it already.
- Decide how to render the Avticate / Deactivate of the connector. Depending how this goes, also find a way to hide the preference completely from rendering if the user is not Premium. `.setEnable(false)` only makes it render greyed out. Somehow we need to to a `.remove()`

Lineflyer commented:
retest this please

fm-sys commented:
> Come up with agreed upon terminology for everything related to ALC

If as subsetting of Geocaching.com, I would call the setting "Adventure Lab" or "show adventure lab caches" (first one preferred)

> also find a way to hide the preference completely from rendering if the user is not Premium. .setEnable(false) only makes it render greyed out. Somehow we need to to a .remove()

I think that could be done in a second step and I would offer you to do (or at least try) that for you :)

> Shorten the geocode in ListViews and on the Popup that shows up when a dot on the map is clicked. So far I have been unable to figure out where exactly the data binding is done otherwise I would have fixed it already.

I think we shouldn't include any connector speciffic code into activitys directly. So rather create a `Geocache.getUserDisplayableGeocode()` method which implements a connector specific interface. But I can do this as well, if you like. It shouldn't be a blocker for merge...

fm-sys reviewed (approved):
The code looks very promising! 

I would really like to see some unit tests for the LC connector. But if you don't feel confident about it yet, I think we can postpone it and include it in the "LC fixup PR" (possibly by me).

I think you have suffered enough (oh dear, 75 (!) comments in this just 4 days old PR). It's probably much easier if we create several small PR's for fixing the remaining issues.

There are some [lint warnings](https://ci.cgeo.org/job/cgeo%20pull%20request/5390/androidLintResult/new/), which make the build failing. If you could fix those I would really like to merge this PR (If none of the team members stops me ;-) )

baiti commented:
> > Come up with agreed upon terminology for everything related to ALC
> 
> If as subsetting of Geocaching.com, I would call the setting "Adventure Lab" or "show adventure lab caches" (first one preferred). The cache type itself maybe "Lab Cache" but "Adventure Lab Cache" is fine as well?

It is not just about how we call the setting. It is  a system wide question. How do we name the "type" of the entity? Check the PR and search for "Adventure::

tools:text="Start Adventure Lab"
<string name="advlab">Adventure Lab Cache</string>
<string name="settings_lc_description">Include Adventure Lab caches on the map.</string>
<string name="cache_alc_start">Open Adventure Lab to start playing this adventure</string>
<string name="cache_alc_install">Install Adventure Lab player to play this adventure</string>
<string name="helper_alc_description">Adventure Lab Player app for your phone or tablet.</string>
<string translatable="false" name="helper_alc">Adventure Lab</string>
ADVLAB("advlab", "Adventure Lab Cache", "", R.string.advlab, R.drawable.type_advlab, "", R.drawable.dot_advlab)

> 
> > also find a way to hide the preference completely from rendering if the user is not Premium. .setEnable(false) only makes it render greyed out. Somehow we need to to a .remove()
> 
> I think that could be done in a second step and I would offer you to do (or at least try) that for you :)
>

Yes. We can work together on this. I want to learn these things.
 
> > Shorten the geocode in ListViews and on the Popup that shows up when a dot on the map is clicked. So far I have been unable to figure out where exactly the data binding is done otherwise I would have fixed it already.
> 
> I think we shouldn't include any connector speciffic code into activitys directly. So rather create a `Geocache.getUserDisplayableGeocode()` method which implements a connector specific interface. But I can do this as well, if you like. It shouldn't be a blocker for merge...

Agree. Let'ts do together.

fm-sys commented:
> There are some lint warnings, which make the build failing. 

Sorry, maybe lint isn't even the problem and everything is fine...

retest this please

fm-sys commented:
> We can work together on this. I want to learn these things.

Probably it's a matter of just a handful of lines. When I started with c:geo / android coding, it always helped me a lot to look at the PR's of others (especially if they were not too extensive). If it's ok for you, I would suggest that when your PR is merged, I'll create two small PR's and ask for your review ;-)

BTW, hooray! Everything is green now... 
![grafik](https://user-images.githubusercontent.com/64581222/114907807-bf8e1f00-9e1b-11eb-9c14-457a7e7b07fc.png)
Thanks for all your work you put into this PR!

baiti commented:
> > Come up with agreed upon terminology for everything related to ALC
> 
> If as subsetting of Geocaching.com, I would call the setting "Adventure Lab" or "show adventure lab caches" (first one preferred). The cache type itself maybe "Lab Cache" but "Adventure Lab Cache" is fine as well?
> 
> > also find a way to hide the preference completely from rendering if the user is not Premium. .setEnable(false) only makes it render greyed out. Somehow we need to to a .remove()
> 
> I think that could be done in a second step and I would offer you to do (or at least try) that for you :)
> 
> > Shorten the geocode in ListViews and on the Popup that shows up when a dot on the map is clicked. So far I have been unable to figure out where exactly the data binding is done otherwise I would have fixed it already.
> 
> I think we shouldn't include any connector speciffic code into activitys directly. So rather create a `Geocache.getUserDisplayableGeocode()` method which implements a connector specific interface. But I can do this as well, if you like. It shouldn't be a blocker for merge...

On my local repo I had already used the following when I made the screenshot in the morning:

```
main/src/cgeo/geocaching/models/Geocache.java:

    @NonNull
    public String getGeocodeShort() {
        final int maxLength = (geocode.length() < 10) ? geocode.length() : 10;
        return geocode.substring(0, maxLength);
    }

```

baiti commented:
> > We can work together on this. I want to learn these things.
> 
> Probably it's a matter of just a handful of lines. When I started with c:geo / android coding, it always helped me a lot to look at the PR's of others (especially if they were not too extensive). If it's ok for you, I would suggest that when your PR is merged, I'll create two small PR's and ask for your review ;-)

Ok.
But let  me explain what my struggle was with the subject of "geocode" all the time. Here is one example. There are many more like this:

```
main/src/cgeo/geocaching/CacheDetailActivity.java:1173

addContextMenu(details.add(R.string.cache_geocode, cache.getGeocode()).right);
```

`addContextMenu` can't be just modified like this:

`addContextMenu(details.add(R.string.cache_geocode, cache.getShortGeocode()).right)`

as it needs the **full** geocode to do its work. Since it does not get a reference to the `cache` object, it cannot do a `cache.getShortGeocode()` at the place where it is needed to bind the string to the ui property.

Of course I could have added another parameter to addContextMenu and pass the `cache.getShortGeocode()` in addition but that can't really be the solution or can it?

> Thanks for all your work you put into this PR!

It was fun ;)

fm-sys commented:
>But let  me explain what my struggle was with the subject of "geocode" all the time. Here is one example. There are many more like this:
>
> ```
>main/src/cgeo/geocaching/CacheDetailActivity.java:1173
>
>addContextMenu(details.add(R.string.cache_geocode, cache.getGeocode()).right);
>```
>
> `addContextMenu` can't be just modified like this:
>
> `addContextMenu(details.add(R.string.cache_geocode, cache.getShortGeocode()).right)`
>
> as it needs the **full** geocode to do its work. Since it does not get a reference to the `cache` object, it cannot do a `cache.getShortGeocode()` at the place where it is needed to bind the string to the ui property.
>
> Of course I could have added another parameter to addContextMenu and pass the `cache.getShortGeocode()` in addition but that can't really be the solution or can it?

Ok, it really seems to be not that easy as I would have expected it. Need to think about it...

> I remember there was a reference code in the beginning, which the user could even define hisself.
E.g. for my ALC I used VelAdventure01 and the link to https://labs.geocaching.com/goto/VelAdventure01 is still working.

Just asking, is this identifier available in theory (even if it may not be usable)? Or the short link like `https://adventurelab.page.link/KSQT`? 

But probably it's better to continue in a dedicated issue...

fm-sys commented:
Opened  #10387 for it...

Lineflyer commented:
> > There are some lint warnings, which make the build failing.
> Sorry, maybe lint isn't even the problem and everything is fine...

Well. those lint warnings look valid and should be solved IMHO: https://ci.cgeo.org/job/cgeo%20pull%20request/5392/androidLintResult/new

Lineflyer reviewed (approved):
Can't judge the code but its implemented in a nice way and working.
As commented:
Well. those lint warnings look valid and should be solved (or commented if not solvable) IMHO: https://ci.cgeo.org/job/cgeo%20pull%20request/5392/androidLintResult/new

baiti commented:
> Can't judge the code but its implemented in a nice way and working.
> As commented:
> Well. those lint warnings look valid and should be solved (or commented if not solvable) IMHO: https://ci.cgeo.org/job/cgeo%20pull%20request/5392/androidLintResult/new

Ok ... .Will remove the unused keys and strings from the XML  and also the one Warning in LCApi.

I don't know what the warning for the .png means.

baiti commented:
Lint complained about unused things .. which I have removed. Now it doesn't compile anymore. I have to go back out the latest changes. I don't want to screw around with git tonight. Will do all this tomorrow. By end of tomorrow it should be all clean and codacy happy and lint too.

fm-sys commented:
What's the state of this PR, can we merge it to master?

baiti commented:
> What's the state of this PR, can we merge it to master?

In general, I'd say yes. I have tested the functionality manually on the emulator and it does what we said it should do.

There is, however, something that should be addressed in the next iteration:. If I set the filter on the main page to just Tradis, then press "Nearest", I get exactly 20 Tradis and at the bottom it says "No more caches". --- Which is, I believe, how it should be.

Now if I set the filter to just ALCs, the list says "7", and offers me a "Load more caches". If I hit that button, I get cache types other than just ALCs. I don't know why the filter isn't working in this case.

Should we try to understand and fix this before it hits the road for the first time?

fm-sys commented:
I would say no. Probably an unrelated issue. It's never a good idea trying to fix to much things in one step. The smaller a PR is, the better it can be reviewed. But we need to create issues for those findings - otherwise they most likely get lost

Lineflyer commented:
Reading the posts above and seeing we have two approvals already, shall we merge?

Afterwards I will also test and report bugs or improvement proposals as dedicated issues for better tracking.
FYI: I created an issue label "Connector LC/ALC" for future use.

baiti commented:
> Reading the posts above and seeing we have two approvals already, shall we merge?

You aren't addressing this to me, right? This goes to the reviewers I'd assume.

From my perspective, yes, go ahead. I've been using the developer build with this PR in it for a week now and haven't noticed any side effects. After all, the interferences with the rest of the system are very shallow, which was a prime objective when I coded it.

baiti commented:
refreshed my local environment:
```
git switch master
git fetch upstream master
git switch minium_ALC
git rebase upstream/master
git push -f
```

jakedot commented:
did the recent changes on geocaching.com affect this PR? I installed the build apk and get no AL Caches anywhere in the app (map, near caches, search) with the Adventure Lab Caches or without.

moving-bits commented:
Sorry for coming back to this PR so late, have been a bit occupied the last days by the recent website changes... :-(

Let's merge this PR to get it into the nightlies and gather some feedback. I'm really looking forward to it!

baiti commented:
@moving-bits ... cool, thanks much. Now we can go ahead and improve it in the areas already discussed, for instance dealing with the ugly id and also come up with a consistent naming of everything we show at the UI.
